### PR TITLE
Specialized struct generics rather than interface

### PR DESCRIPTION
**Pull request by benaadams**
Use struct wrappers to specialize generic dispatch

`Generic<object> Dispatch` < `Interface Dispatch` < `Virtual Dispatch` < `Generic<struct> Dispatch`

Follow up to https://github.com/aspnet/KestrelHttpServer/pull/1624#issuecomment-292521621

### CODE REVIEW

**Comment by davidfowl** [comment]
This code is too ugly, the improvement needs to be worth it. You should also be able to do better if you did:

```C#
public interface IHttpParser<TRequestHandler> where TRequestHandler struct,
                                                                                 where TRequestHandler IHttpHeadersHandler,
                                                                                 where TRequestHandler IHttpRequestLineHandler,
{
}
```

**Comment by benaadams** [comment]
Pre
```
                                   Method |          Mean |     StdDev |       Op/s | Scaled | Allocated |
----------------------------------------- |-------------- |----------- |----------- |------- |---------- |
                     PlaintextTechEmpower |   982.1983 ns | 11.5242 ns | 1018124.33 |   1.00 |     349 B |
                     PlaintextAbsoluteUri | 1,596.0722 ns | 17.9466 ns |  626538.05 |   1.63 |     573 B |
            PipelinedPlaintextTechEmpower |   862.9871 ns |  7.9130 ns | 1158765.81 |   0.88 |     348 B |
 PipelinedPlaintextTechEmpowerDrainBuffer |   747.2833 ns | 10.7552 ns | 1338180.52 |   0.76 |     348 B |
                               LiveAspNet | 1,989.2925 ns | 15.6089 ns |  502691.29 |   2.03 |   1.07 kB |
                      PipelinedLiveAspNet | 2,034.8853 ns | 23.7429 ns |  491428.19 |   2.07 |   1.07 kB |
                                  Unicode | 3,374.3776 ns | 42.9808 ns |  296350.95 |   3.44 |   1.85 kB |
                         UnicodePipelined | 3,394.8923 ns | 21.8284 ns |  294560.15 |   3.46 |   1.86 kB |
```
Post
```
                                   Method |          Mean |     StdDev |       Op/s | Scaled | Allocated |
----------------------------------------- |-------------- |----------- |----------- |------- |---------- |
                     PlaintextTechEmpower |   982.0322 ns |  6.1319 ns | 1018296.51 |   1.00 |     373 B |
                     PlaintextAbsoluteUri | 1,571.6325 ns | 17.6337 ns |  636281.08 |   1.60 |     597 B |
            PipelinedPlaintextTechEmpower |   872.5876 ns | 13.4669 ns | 1146016.76 |   0.89 |     372 B |
 PipelinedPlaintextTechEmpowerDrainBuffer |   741.8345 ns |  6.0405 ns | 1348009.60 |   0.76 |     372 B |
                               LiveAspNet | 1,921.2996 ns | 13.8283 ns |  520481.02 |   1.96 |   1.09 kB |
                      PipelinedLiveAspNet | 1,895.9246 ns | 14.0616 ns |  527447.13 |   1.93 |    1.1 kB |
                                  Unicode | 3,241.2465 ns | 29.0088 ns |  308523.28 |   3.30 |   1.88 kB |
                         UnicodePipelined | 3,291.7754 ns | 17.9002 ns |  303787.43 |   3.35 |   1.89 kB |
```

/cc @davidfowl

**Comment by benaadams** [comment]
Missed a function taking an interface also

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/ServiceContext.cs [hunk 1]
```diff
@@ -7,13 +7,13 @@
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
 {
-    public class ServiceContext
+    public class ServiceContext<TRequestHandler> where TRequestHandler : struct, IHttpHeadersHandler, IHttpRequestLineHandler
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/ServiceContext.cs]: Why is this generic? Just make it have the func be `Func<FrameAdapter, IHttpParser<FrameAdapter>>`

**Comment by benaadams** [comment]
```
                                   Method |          Mean |      StdDev |       Op/s | Scaled | Allocated |
----------------------------------------- |-------------- |------------ |----------- |------- |---------- |
                     PlaintextTechEmpower |   958.3939 ns |  17.5647 ns |  1043412.3 |   1.00 |     349 B |
                     PlaintextAbsoluteUri | 1,601.9326 ns | 114.1761 ns |     624246 |   1.67 |     573 B |
            PipelinedPlaintextTechEmpower |   831.2717 ns |   7.0609 ns | 1202976.18 |   0.87 |     348 B |
 PipelinedPlaintextTechEmpowerDrainBuffer |   718.7616 ns |   5.7565 ns | 1391281.85 |   0.75 |     348 B |
                               LiveAspNet | 1,870.5568 ns |  14.5679 ns |  534600.17 |   1.95 |   1.07 kB |
                      PipelinedLiveAspNet | 1,862.9649 ns |  13.5595 ns |  536778.77 |   1.94 |   1.07 kB |
                                  Unicode | 3,256.7373 ns |  43.1425 ns |  307055.77 |   3.40 |   1.85 kB |
                         UnicodePipelined | 3,276.1174 ns |  21.1888 ns |  305239.37 |   3.42 |   1.86 kB |
```

**Comment by benaadams** [comment]
Pre
```
                        Method |          Mean |     StdDev |       Op/s | Scaled | Allocated |
------------------------------ |-------------- |----------- |----------- |------- |---------- |
          PlaintextTechEmpower |   978.9098 ns | 13.6899 ns | 1021544.61 |   1.00 |     349 B |
          PlaintextAbsoluteUri | 1,576.9072 ns | 17.3851 ns |  634152.71 |   1.61 |     573 B |
 PipelinedPlaintextTechEmpower |   847.7863 ns |  9.8449 ns | 1179542.58 |   0.87 |     348 B |
            Pipe...DrainBuffer |   738.4561 ns | 16.1327 ns | 1354176.55 |   0.75 |     348 B |
                    LiveAspNet | 1,955.0006 ns | 60.4765 ns |  511508.79 |   2.00 |   1.07 kB |
           PipelinedLiveAspNet | 1,982.8594 ns | 10.9114 ns |   504322.2 |   2.03 |   1.07 kB |
                       Unicode | 3,291.3795 ns | 96.0813 ns |  303823.97 |   3.36 |   1.85 kB |
              UnicodePipelined | 3,335.3001 ns | 14.9179 ns |  299823.09 |   3.41 |   1.86 kB |
```
Post
```
                        Method |          Mean |     StdDev |       Op/s | Scaled | Allocated |
------------------------------ |-------------- |----------- |----------- |------- |---------- |
          PlaintextTechEmpower |   942.4937 ns |  6.8881 ns | 1061015.08 |   1.00 |     349 B |
          PlaintextAbsoluteUri | 1,584.2315 ns | 44.3674 ns |  631220.89 |   1.68 |     573 B |
 PipelinedPlaintextTechEmpower |   830.6567 ns | 10.1710 ns | 1203866.73 |   0.88 |     348 B |
            Pipe...DrainBuffer |   737.4760 ns | 32.7704 ns |  1355976.4 |   0.78 |     348 B |
                    LiveAspNet | 1,916.9054 ns | 39.2063 ns |  521674.15 |   2.03 |   1.07 kB |
           PipelinedLiveAspNet | 1,894.0680 ns | 20.3642 ns |  527964.16 |   2.01 |   1.07 kB |
                       Unicode | 3,265.3568 ns | 49.3643 ns |  306245.25 |   3.46 |   1.85 kB |
              UnicodePipelined | 3,270.2079 ns | 18.1554 ns |  305790.96 |   3.47 |   1.86 kB |
```

**Comment by davidfowl** [comment]
FYI @vancem @jkotas

**Comment by davidfowl** [comment]
That's a pretty nice boost.

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/ISocketOutput.cs [hunk 2]
```diff
@@ -17,6 +17,6 @@ public interface ISocketOutput
         Task WriteAsync(ArraySegment<byte> buffer, bool chunk = false, CancellationToken cancellationToken = default(CancellationToken));
         void Flush();
         Task FlushAsync(CancellationToken cancellationToken = default(CancellationToken));
-        void Write<T>(Action<WritableBuffer, T> write, T state);
+        void Write<T>(Action<WritableBuffer, T> write, T state) where T : struct;
```
Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/ISocketOutput.cs]: Should this be `ISocketOutput<T> where T : struct`?

Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/ISocketOutput.cs]: I guess so yeah

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/HttpParser.cs [hunk 3]
```diff
@@ -9,7 +9,7 @@
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http
 {
-    public class HttpParser : IHttpParser
+    public class HttpParser<TRequestHandler> : IHttpParser<TRequestHandler> where TRequestHandler : struct, IHttpHeadersHandler, IHttpRequestLineHandler
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/HttpParser.cs]: Can we remove the where `TRequestHandler` struct? This is a performance optimization not a correctness thing. It means the test code and benchmarks can avoid those little struct wrappers.

Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/HttpParser.cs]: Sure

**Comment by davidfowl** [comment]
@benaadams Can you fix the conflicts?

**Comment by benaadams** [comment]
Rebased

**Comment by benaadams** [comment]
> Type system abuse LGTM.

I like to call it "Working with the type system" 😉

**Comment by molinch** [comment]
Stupid question, but couldn't the C# compiler automatically generate 2 methods when we have a generic type without constraint?
Let's say we have `Method<T>(T something)`

If the compiler generated both:
`Method<T>(T something) where T:struct`
`Method<T>(T something) where T:class`

We wouldn't need to 'hack' the code.

**Comment by benaadams** [comment]
Compiler generates a single method for all class types (single size `IntPtr`; includes method table); and a method for every struct type (different sizes; method table not included in struct).

So it does convert
```csharp
Method<T>(T something)
```
into
```csharp
// One for all classes
Method<T>(T something) where T:class 
// One for each struct
Method<T>(T something) where T:int
Method<T>(T something) where T:char
Method<T>(T something) where T:long
Method<T>(T something) where T:MyStruct
Method<T>(T something) where T:YourStruct
```

The "hack" is to call it with a struct type rather than a class type

**Comment by molinch** [comment]
Ok it's clear, sorry I misunderstood it.

Basically you added generics so it will generate a specific method for it, having the right type, so no callvirt anymore and no boxing since it won't call by passing the interface.

Is it right?

**Comment by benaadams** [comment]
Yep :)

**Pull request merged**